% Generated by roxygen2 (4.1.0.9000): do not edit by hand
% Please edit documentation in R/database.R
\name{import.marker.set}
\alias{import.marker.set}
\title{Import marker data.}
\usage{
import.marker.set(name, markers, db, description = NA, genome.id = NULL,
  start.transaction = TRUE)
}
\arguments{
\item{name}{Name for the marker set.}

\item{markers}{vector of marker IDs or path to marker data file or data.frame containing marker
information. See Details.}

\item{db}{the datbase connection.}

\item{description}{Text description of the marker set.}

\item{genome.id}{ID of the genome build from which the markers were sourced.}

\item{start.transaction}{set to false if calling from another method that starts a transaction.}
}
\value{
If markers is a path, returns a list with two elements, set.id and markers. Otherwise returns the
integer ID of the new marker set.
}
\description{
Currently, CLASP only supports bi-allelic markers and does not support sex chromosome markers.
The later is because cell lines are often of indeterminate sex or exhibit sex chromosome mosaicism.
}
\details{
If \code{markers} is a vector, the elements are marker IDs and the names can optionally be marker
set-specific names. If it is a data.frame, it must have five columns:
1. name: Name of the marker.
2. chromosome: Name of the chromosome. Must be an integer.
3. bp: Physical position of the marker.
4. cm: Genetic position of the marker (in centimorgans).
5. alleles: String listing the two possible genotypes for this marker. The alleles should be listed
with the reference allele first, followed by the variant allele. If this marker is to be
associated with intensity data, the intensity values should be reordered to match the allele order.
}

